From 72a509efefdea3297ef0d3c17265c5937c26843b Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Tue, 16 Sep 2014 20:20:18 -0400
Subject: [fs] nfsd: Ensure lookup_clientid() takes client_lock

Message-id: <1410898849-22126-135-git-send-email-bfields@redhat.com>
Patchwork-id: 95773
O-Subject: [Patch RHEL7 134/165] nfsd: Ensure lookup_clientid() takes client_lock
Bugzilla: 1044798 1123721 1141932
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Trond Myklebust <trond.myklebust@primarydata.com>

Ensure that the client lookup is done safely under the client_lock, so
we're not relying on the client_mutex.

Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: 3e339f964b74b7223ab128f36f4b2aaf9dc12eb9
Bugzilla: 1044798
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4state.c b/fs/nfsd/nfs4state.c
index 68383b0..f9d077d 100644
--- a/fs/nfsd/nfs4state.c
+++ b/fs/nfsd/nfs4state.c
@@ -3451,13 +3451,17 @@ static __be32 lookup_clientid(clientid_t *clid,
   * will be false.
   */
  WARN_ON_ONCE(cstate->session);
+ spin_lock(&nn->client_lock);
  found = find_confirmed_client(clid, false, nn);
- if (!found)
+ if (!found) {
+  spin_unlock(&nn->client_lock);
   return nfserr_expired;
+ }
+ atomic_inc(&found->cl_refcount);
+ spin_unlock(&nn->client_lock);
 
  /* Cache the nfs4_client in cstate! */
  cstate->clp = found;
- atomic_inc(&found->cl_refcount);
  return nfs_ok;
 }
 
-- 
1.7.1